### PR TITLE
feat(zk): implement `SingleChipLayouter`

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements [floor_planner/single_pass](https://github.com/kroma-network/halo2/blob/dev/halo2_proofs/src/circuit/floor_planner/single_pass.rs) which is an implementation of `FloorPlanner` based on #118 

`SimpleTableLayouter` and `SingleChipLayouter` are not clear to me so refactoring is not done at this moment.

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Comment by TomTaehoonKim** [comment]
@chokobole Could you fix the commit message of 1d7c1b16 to "fix missing changes from ~"?

File: tachyon/zk/plonk/circuit/region.h [hunk 1]
```diff
@@ -98,6 +98,91 @@ class Region {
...
+  explicit Region(Layouter* layouter) : layouter_(layouter) {}
+
+  // See the comment above.
+  Error EnableSelector(AnnotateCallback annotate, const Selector& selector,
+                       size_t offset) {
+    return layouter_->EnableSelector(std::move(annotate), selector, offset);
+  }
+
+  // See the comment above.
+  void NameColumn(AnnotateCallback annotate, const AnyColumn& column) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:111]:
```suggestion
  void NameColumn(AnnotateCallback annotate, const AnyColumn& column) override {
```
Plus, if we get to add override keyword, how about removing " // See the comment above." comment?

File: tachyon/zk/plonk/circuit/region.h [hunk 2]
```diff
@@ -98,6 +98,91 @@ class Region {
...
+    if (error != Error::kNone) return error;
+    *assigned_cell = {cell, value};
+    return Error::kNone;
+  }
+
+  // See the comment above.
+  Error AssignAdviceFromConstant(AnnotateCallback annotate,
+                                 const AdviceColumn& column, size_t offset,
+                                 const F& constant,
+                                 AssignedCell<F>* assigned_cell) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:137]:
```suggestion
                                 AssignedCell<F>* assigned_cell) override {
```
Plus, if we get to add override keyword, how about removing " // See the comment above." comment?

File: tachyon/zk/plonk/circuit/region.h [hunk 3]
```diff
@@ -98,6 +98,91 @@ class Region {
...
+
+  // See the comment above.
+  void NameColumn(AnnotateCallback annotate, const AnyColumn& column) {
+    layouter_->NameColumn(std::move(annotate), column);
+  }
+
+  // See the comment above.
+  Error AssignAdvice(AnnotateCallback annotate, const AdviceColumn& column,
+                     size_t offset, AssignCallback assign,
+                     AssignedCell<F>* assigned_cell) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:118]:
```suggestion
                     AssignedCell<F>* assigned_cell) override {
```
Plus, if we get to add override keyword, how about removing " // See the comment above." comment?

File: tachyon/zk/plonk/circuit/region.h [hunk 4]
```diff
@@ -98,6 +98,91 @@ class Region {
       return Error::kNone;
     }
   };
+
+  explicit Region(Layouter* layouter) : layouter_(layouter) {}
+
+  // See the comment above.
+  Error EnableSelector(AnnotateCallback annotate, const Selector& selector,
+                       size_t offset) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:106]:
```suggestion
                       size_t offset) override {
```
Plus, if we get to add override keyword, how about removing " // See the comment above." comment?

File: tachyon/zk/plonk/circuit/region.h [hunk 5]
```diff
@@ -98,6 +98,91 @@ class Region {
...
+    if (error != Error::kNone) return error;
+    *assigned_cell = {cell, Value<F>::Known(constant)};
+    return Error::kNone;
+  }
+
+  // See the comment above.
+  Error AssignAdviceFromInstance(AnnotateCallback annotate,
+                                 const InstanceColumn& instance, size_t row,
+                                 const AdviceColumn& advice, size_t offset,
+                                 AssignedCell<F>* cell) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:150]:
```suggestion
                                 AssignedCell<F>* cell) override {
```
Plus, if we get to add override keyword, how about removing " // See the comment above." comment?

File: tachyon/zk/plonk/circuit/region.h [hunk 6]
```diff
@@ -98,6 +98,91 @@ class Region {
...
+    return layouter_->AssignAdviceFromInstance(std::move(annotate), instance,
+                                               row, advice, offset, cell);
+  }
+
+  // See the comment above.
+  Error AssignFixed(AnnotateCallback annotate, const FixedColumn& column,
+                    size_t offset, AssignCallback assign,
+                    AssignedCell<F>* assigned_cell) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:158]:
```suggestion
                    AssignedCell<F>* assigned_cell) override {
```
Plus, if we get to add override keyword, how about removing " // See the comment above." comment?

File: tachyon/zk/plonk/circuit/region.h [hunk 7]
```diff
@@ -98,6 +98,91 @@ class Region {
...
+        },
+        &cell);
+    if (error != Error::kNone) return error;
+    *assigned_cell = {cell, value};
+    return Error::kNone;
+  }
+
+  // See the comment above.
+  Error ConstrainConstant(const Cell& cell,
+                          const math::RationalField<F>& constant) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:175]:
```suggestion
                          const math::RationalField<F>& constant) override {                          
```
Plus, if we get to add override keyword, how about removing " // See the comment above." comment?

File: tachyon/zk/plonk/circuit/region.h [hunk 8]
```diff
@@ -98,6 +98,91 @@ class Region {
...
+    return layouter_->ConstrainConstant(cell, constant);
+  }
+
+  // See the comment above.
+  Error ConstrainEqual(const Cell& left, const Cell& right) {
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:180]:
```suggestion
  Error ConstrainEqual(const Cell& left, const Cell& right) override {
```
Plus, if we get to add override keyword, how about removing " // See the comment above." comment?

File: tachyon/zk/plonk/circuit/region.h [hunk 9]
```diff
@@ -98,6 +98,91 @@ class Region {
...
+    return layouter_->ConstrainEqual(left, right);
+  }
+
+ private:
+  Layouter* const layouter_;
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/region.h:185]: Don't we need a getter for this member?

File: tachyon/zk/plonk/circuit/floor_planner/single_chip_layouter.h [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+      Cell cell;
+      error = AssignAdvice(
+          std::move(annotate), advice, offset,
+          [&value]() { return math::RationalField<F>(value); }, &cell);
+      if (error != Error::kNone) return error;
+
+      error = layouter_->assignment_->Copy(
+          cell.column(),
+          layouter_->regions_[cell.region_index()] + cell.row_offset(),
+          instance, row);
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/floor_planner/single_chip_layouter.h:87]:
```suggestion
          instance, row);
      if (error != Error::kNone) return error;
```

File: tachyon/zk/plonk/circuit/floor_planner/single_chip_layouter.h [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+          right.column(),
+          layouter_->regions_[right.region_index()] + right.row_offset());
+    }
+
+   private:
+    SingleChipLayouter* const layouter_;
+    size_t region_index_ = 0;
+    // Stores the constants to be assigned, and the cells to which they are
+    // copied.
+    Constants constants_;
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/floor_planner/single_chip_layouter.h:123]: Don't we need getters for these members?

File: tachyon/zk/plonk/circuit/floor_planner/single_chip_layouter.h [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+  Error AssignRegion(NameCallback name, AssignRegionCallback assign) override {
+    std::string region_name = name.Run();
+    size_t region_index = regions_.size();
+
+    // Get shape of the region.
+    RegionShape<F> shape(region_index);
+    {
+      // TODO(chokobole): Add event trace using
+      // https://github.com/google/perfetto.
+      VLOG(1) << "Assign region 1st pass: " << region_name;
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/floor_planner/single_chip_layouter.h:153]: Can we just use `VLOG` without importing any headers?

File: tachyon/zk/plonk/circuit/floor_planner/simple_table_layouter.h [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    if (value.assigned.size() <= offset) {
+      value.assigned.resize(offset + 1, false);
+    }
+    value.assigned[offset] = true;
+    return Error::kNone;
+  }
+
+ private:
+  Assignment<F>* const assignment_;
+  std::vector<TableColumn>* const used_columns_;
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/zk/plonk/circuit/floor_planner/simple_table_layouter.h:83]: Nope, but it should be typed with `const`.

File: tachyon/zk/plonk/circuit/region.h (hunk 9 above)
Comment (chokobole) [reply tachyon/zk/plonk/circuit/region.h:185]: Nope

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Pull request merged**